site-scrapers/Hannaford.js

```diff
@@ -63,10 +65,77 @@ async function ScrapeWebsiteData(browser) {
                 const filename =
                     "hannaford-zip-" + loc.zip + "-date-" + today + ".png";
                 await page.screenshot({ path: filename });
+                await saveBodyHtml(page);
             }
             results[loc.zip] = result;
         }
     }
 
     return results;
 }
+
+async function writeFile(filename, data) {
+    return new Promise((resolve, reject) => {
+        fs.writeFile(filename, data, (err) => {
+            if (err) {
+                reject(err);
+            } else {
+                resolve();
+            }
+        });
+    });
+}
+
+async function saveBodyHtml(page) {
+    // It appears puppeteer does not a generic page.waitForNetworkIdle
+    // so for now we just assume all XHR requests will complete within
+    // a reasonable timeframe, for now 10s.
+    await new Promise((resolve) => {
+        setTimeout(function () {
+            resolve("done waiting");
+        }, 1000);
+    });
+
+    const body = await page.$("body");
+    const html = await body.evaluate((node) => node.outerHTML);
+    const timestamp =
+        new Date().toISOString().substring(0, 16).replace(":", "") + "Z";
+    const fileName = ["hannaford-", timestamp, ".html"].join("");
+    await writeFile(fileName, html);
+    await uploadFileToS3("debug", fileName);
+    return fileName;
+}
+
+async function uploadFileToS3(bucketDir, fileName) {
+    const fileContents = await new Promise((resolve, reject) => {
+        fs.readFile(fileName, "utf-8", async (err, data) => {
+            if (err) {
+                reject(err);
+            } else {
+                resolve(data);
+            }
+        });
+    });
+
+    const s3 = new AWS.S3({
```

Comment:
could we make a utility function to upload to our S3 bucket, and then just use that in this file and main.js? (And StopAndShop I'm guessing)

Reply:
Defo.